### ISSUE TITLE
Memory and CPU Issue

### ISSUE THREAD

**Issue by holiday-sunrise** [avioconsulting/mule-opentelemetry-module#197]
We have an API with 100 thousand reguest per 30 min. 

With the Module the Memory Consumption 3 time higher and the time per request slows down from 1-4 ms  to 10-70.

10 times slower :-(

what are we doing wrong

Mule 4.6.6
Java 8
Rest API as Anti Corruption Layer for LDAP
opensuse 15.6
2 CPU max per replica
2.6 GB Memory max  per replica

![image](https://github.com/user-attachments/assets/bc36d738-9b62-4716-a988-8cfdd6efebbc)

**Comment by manikmagar** [comment]
@holiday-sunrise Do you have usage of `http:request` in those flows? If yes, see if this helps - https://avioconsulting.github.io/mule-opentelemetry-module/#_static_vs_dynamic_global_configurations

**Comment by holiday-sunrise** [comment]
we using the following config

````xml
<opentelemetry:config name="OpenTelemetry_Config_Elk" serviceName="${api.name}" spanAllProcessors="true">
		<opentelemetry:resource-attributes>
			<opentelemetry:attribute key="mule.env" value="${mule.env}" />
			<opentelemetry:attribute key="deployment.environment" value="${mule.env}" />
		</opentelemetry:resource-attributes>
		<opentelemetry:exporter>
			<opentelemetry:otlp-exporter collectorEndpoint="${ELASTIC_APM_SERVER_URL}">
				<opentelemetry:headers>
					<opentelemetry:header key="Authorization" value="Bearer ${ELASTIC_APM_SECRET_TOKEN}" />
				</opentelemetry:headers>
			</opentelemetry:otlp-exporter>
		</opentelemetry:exporter>
	</opentelemetry:config>
````

**Comment by holiday-sunrise** [comment]
> @holiday-sunrise Do you have usage of `http:request` in those flows? If yes, see if this helps - https://avioconsulting.github.io/mule-opentelemetry-module/#_static_vs_dynamic_global_configurations

is this decreasing the throughput ?

**Comment by holiday-sunrise** [comment]
But the performance is worse 10 times slower ? I think there is still a config problem. Can we reduse spans? Like st variable oder cal custom module ? 

Did you that will increase the performance ?

**Comment by manikmagar** [comment]
Also, check for the static vs. dynamic usage. You have shared OTEL config but not the HTTP request config earlier.

> @holiday-sunrise Do you have usage of `http:request` in those flows? If yes, see if this helps - https://avioconsulting.github.io/mule-opentelemetry-module/#_static_vs_dynamic_global_configurations

**Comment by holiday-sunrise** [comment]
We will check it

**Comment by holiday-sunrise** [comment]
> Do you have usage of `http:request` in those flows? If yes, see if this helps -

Yes we habe but not always sometimes its an custom connector (module)

**Comment by holiday-sunrise** [comment]
> (eg. mule:set-variable).

is there a list of all configs like mule:set-variable

Did you have an example

**Comment by holiday-sunrise** [comment]
@manikmagar any News ? My you help us ? 

What is a best practice config with a limit cunsumtion overhead

**Comment by manikmagar** [comment]
Hi @holiday-sunrise, we are also doing some tests. If you would like to perform the similar test, I suggest to try this - 

- Use the latest `2.3.0-SNAPSHOT`
- Disable the Interceptor feature by setting property `mule.otel.interceptor.processor.enable=false` on runtime
- Without interceptor this is needed for context propagation, add the following operation at the beginning of the main flow (e.g. After the HTTP-listener in APIKit flow). [If you were using `opentelemetry:get-trace-context` for anything, replace it with this one]
```
<opentelemetry:get-current-trace-context doc:name="Get Current Trace Context" config-ref="OpenTelemetry_Config" target="OTEL_TRACE_CONTEXT"/>
```
- Verify the traces generated and the performance.

To use the snapshot version, you would need to add the following repository to the pom file -

```
<repository>
        <id>oss.sonatype.org-snapshot</id>
        <url>https://oss.sonatype.org/content/repositories/snapshots</url>
        <releases>
            <enabled>false</enabled>
        </releases>
        <snapshots>
            <enabled>true</enabled>
        </snapshots>
    </repository>
```

If you could please share your observations here, that will help us too. Thank you!

**Comment by holiday-sunrise** [comment]
> Hi @holiday-sunrise, we are also doing some tests. If you would like to perform the similar test, I suggest to try this -
> 
> * Use the latest `2.3.0-SNAPSHOT`
> * Disable the Interceptor feature by setting property `mule.otel.interceptor.processor.enable=false` on runtime
> * Without interceptor this is needed for context propagation, add the following operation at the beginning of the main flow (e.g. After the HTTP-listener in APIKit flow). [If you were using `opentelemetry:get-trace-context` for anything, replace it with this one]
> 
> ```
> <opentelemetry:get-current-trace-context doc:name="Get Current Trace Context" config-ref="OpenTelemetry_Config" target="OTEL_TRACE_CONTEXT"/>
> ```
> 
> * Verify the traces generated and the performance.
> 
> To use the snapshot version, you would need to add the following repository to the pom file -
> 
> ```
> <repository>
>         <id>oss.sonatype.org-snapshot</id>
>         <url>https://oss.sonatype.org/content/repositories/snapshots</url>
>         <releases>
>             <enabled>false</enabled>
>         </releases>
>         <snapshots>
>             <enabled>true</enabled>
>         </snapshots>
>     </repository>
> ```
> 
> If you could please share your observations here, that will help us too. Thank you!

In which format do you need the observations

**Comment by manikmagar** [comment]
At least a summary similar to what you shared in the original question would do. Basically, with those suggested changes, how does your app behave when compared to the original? If there are CPU and Memory graphs, that would be addon. Thank you!

**Comment by KonnoroYatsu** [comment]
> * Use the latest `2.3.0-SNAPSHOT`
> * Disable the Interceptor feature by setting property `mule.otel.interceptor.processor.enable=false` on runtime
> * Without interceptor this is needed for context propagation, add the following operation at the beginning of the main flow (e.g. After the HTTP-listener in APIKit flow). [If you were using `opentelemetry:get-trace-context` for anything, replace it with this one]
> 
> ```
> <opentelemetry:get-current-trace-context doc:name="Get Current Trace Context" config-ref="OpenTelemetry_Config" target="OTEL_TRACE_CONTEXT"/>
> ```
> 
> * Verify the traces generated and the performance.
> 
> To use the snapshot version, you would need to add the following repository to the pom file -
> 
> ```
> <repository>
>         <id>oss.sonatype.org-snapshot</id>
>         <url>https://oss.sonatype.org/content/repositories/snapshots</url>
>         <releases>
>             <enabled>false</enabled>
>         </releases>
>         <snapshots>
>             <enabled>true</enabled>
>         </snapshots>
>     </repository>
> ```
> 
> If you could please share your observations here, that will help us too. Thank you!

@manikmagar I did some tests using your recommendation and got the following results:
**Application before adding the OpenTelemetry module:**
<img width="958" alt="image_01" src="https://github.com/user-attachments/assets/953d622c-5b04-40e9-ac40-05d1ddfd5c52">

**Application using module version 2.2.1 (The application broke the next day):**
<img width="960" alt="image_02" src="https://github.com/user-attachments/assets/663e2f57-ff22-4e8b-b72c-52a1fe254ace">

**Application using module version 2.3.0-SNAPSHOT, with mule.otel.interceptor.processor.enable=false and opentelemetry:get-current-trace-context after each listener:**
<img width="960" alt="image_03" src="https://github.com/user-attachments/assets/6ba355fc-bb09-4ab2-8c4a-60e1fbe7032b">

Your recommendation worked for me. When do you plan to release version 2.3.0?

**Comment by adesjardin** [comment]
We expect to have 2.3.0 released later this week or early next week.  We should have a quick follow up release after that to try and resolve some of the functionality that is removed if interceptors are turned off.

**Comment by manikmagar** [comment]
@KonnoroYatsu The 2.3.0 was released today.

**Comment by KonnoroYatsu** [comment]
> @KonnoroYatsu The 2.3.0 was released today.

Nice. Thank you!

**Comment by holiday-sunrise** [comment]
Was Means that. Without Tracing ID it makes no sense, or!

**Comment by holiday-sunrise** [comment]
> If you could please share your observations here, that will help us too. Thank you!
how can i do that

**Comment by manikmagar** [comment]
Hi @holiday-sunrise,

The required modifications are described here - https://github.com/avioconsulting/mule-opentelemetry-module/issues/197#issuecomment-2327976890.

We have already received some results here: https://github.com/avioconsulting/mule-opentelemetry-module/issues/197#issuecomment-2355678169. The Snapshot mentioned was released a while back (the latest version is 2.6.0). We appreciate your input so far on this. 

But we will close this issue as resolved. If you observe anything new, please feel free to open a new issue.

**Comment by holiday-sunrise** [comment]
thx it works fine. soory for the late feedback on your fix
